### PR TITLE
Support snapshot/checkpoint/backup for string and sorted data

### PR DESCRIPTION
**Pull request by JiayuZzz**
### What problem does this PR solve?

1. Support make a snapshot which capture a immutable view of KVDK instance on the making time
2. Support make a checkpoint of KVDK on a snapshot, the checkpoint can be recovered in re-opening the instance
3. Support make a backup of a KVDK on a snapshot to another path (experimental)

### What is changed and how it works?

What's Changed:

1. Add a old_version_offset area in string and sorted data record which consists of a version chain on PMem
2. Add delete record to maintain version chain of deleted data
3. Old version data records that refered by a snapshot will not be freed until the snapshot is released. The free work is distributed to foreground threads and background thread to amortize the overhead.
4. In recovery, if recover from a checkpoint or a backup instance, the data newer than checkpoint/backup version will be dropped, and the hashtable and the linkage of skiplist will be repaired to the checkpoint/backup version according to the version chain
5. Merge write threads and read threads as access_thread, and share same thread resources
6. Each access_thread will maintain a temporal local snapshot in processing write request to avoid ABA problem 

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Extra DRAM usage for maintain refered old records, the DRAM usage is negligible if no user snapshot holding, otherwise it depends to the lifetime of user snapshot and how many update/delete operations happened during the lifetime
    - ~20% drop on update performance and 30%~40% drop on delete performance for extra cleaning work of old records

### CODE REVIEW

**Comment by JiayuZzz** [comment]
Adding more comments and tests

**Comment by JiayuZzz** [comment]
The multi-version read (read data on a specified snapshot) is implemented but not pushed on this PR, as it involves API changes

File: engine/hash_table.hpp [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #include <vector>
 
 #include "kvdk/engine.hpp"
+#include "kvdk/namespace.hpp"
```
Comment (sherry-1001) [inline engine/hash_table.hpp:13]: is necessary？

File: engine/kv_engine.hpp [hunk 2]
```diff
@@ -92,26 +108,40 @@ class KVEngine : public Engine {
...
+  void ReleaseAccessThread() override { access_thread.Release(); }
 
   const std::vector<std::shared_ptr<Skiplist>> &GetSkiplists() {
     return skiplists_;
   };
 
 private:
+  friend OldRecordsCleaner;
+
+  KVEngine(const Configs &configs)
```
Comment (sherry-1001) [inline engine/kv_engine.hpp:120]: Confused parameters compared with `Init`. Consider only pass `configs.max_access_thread`?

File: engine/kv_engine.cpp [hunk 3]
```diff
@@ -92,42 +90,32 @@ void KVEngine::ReportPMemUsage() {
...
-    if (count_down1 < background_interval) {
-      FreeSkiplistDramNodes();
-      count_down1 = free_skiplist_node_interval;
-    } else {
-      count_down1 -= background_interval;
-    }
+void KVEngine::startBackgroundWorks() {
+  std::unique_lock<SpinMutex> ul(bg_work_signals_.terminating_lock);
+  bg_work_signals_.terminating = false;
+  bg_threads_.emplace_back(&KVEngine::backgroundPMemAllocatorOrgnizer, this);
```
Comment (sherry-1001) [inline engine/kv_engine.cpp:96]: `backgroundPMemAllocatorOrganizer`

File: engine/utils/utils.hpp [hunk 4]
```diff
@@ -36,6 +38,24 @@
 
 namespace KVDK_NAMESPACE {
 
+struct RAIICaller {
```
Comment (sherry-1001) [inline engine/utils/utils.hpp:41]: add comments. By the way `RAII` is not suitable..

File: engine/kv_engine.hpp (hunk 2 above)
Comment (JiayuZzz) [reply engine/kv_engine.hpp:120]: it's a private constructor so I think it's ok

File: engine/hash_table.hpp (hunk 1 above)
Comment (JiayuZzz) [reply engine/hash_table.hpp:13]: no, just for make sure it's not missed

File: engine/kv_engine.cpp [hunk 5]
```diff
@@ -757,13 +849,27 @@ Status KVEngine::RestorePendingBatch() {
...
+  }
+
+  if (s == Status::Ok) {
+    s = MaybeRestoreBackup();
+  }
+
   if (s != Status::Ok) {
     return s;
   }
   GlobalLogger.Info("RestorePendingBatch done.\n");
```
Comment (sherry-1001) [inline engine/kv_engine.cpp:864]: move this log to 852 line

File: engine/kv_engine.cpp [hunk 6]
```diff
@@ -480,16 +477,25 @@ Status KVEngine::RestoreSkiplistHead(DLRecord *pmem_record, const DataEntry &) {
 Status KVEngine::RestoreStringRecord(StringRecord *pmem_record,
                                      const DataEntry &cached_entry) {
   assert(pmem_record->entry.meta.type & StringRecordType);
+  if (string_view_2_string(pmem_record->Key()) == "key1") {
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:480]: What's this? Debug info?

File: engine/kv_engine.cpp [hunk 7]
```diff
@@ -932,16 +1104,20 @@ Status KVEngine::SSetImpl(Skiplist *skiplist, const StringView &user_key,
     HashEntry hash_entry;
     DataEntry data_entry;
     auto hint = hash_table_->GetHint(collection_key);
-    std::lock_guard<SpinMutex> lg(*hint.spin);
-    Status s =
-        hash_table_->SearchForWrite(hint, collection_key, SortedDataRecord,
-                                    &entry_ptr, &hash_entry, &data_entry);
+    std::unique_lock<SpinMutex> ul(*hint.spin);
+    RAIICaller local_snapshot_holder(
+        [&]() { version_controller_.HoldLocalSnapshot(); },
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1109]: Create a token class to perform RAII

File: engine/kv_engine.cpp [hunk 8]
```diff
@@ -1119,30 +1303,34 @@ Status KVEngine::BatchWrite(const WriteBatch &write_batch) {
     return s;
   }
 
+  RAIICaller batch_write_holder(
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1306]: Make this a lock and guard it.

File: engine/kv_engine.cpp [hunk 9]
```diff
@@ -1213,14 +1415,13 @@ Status KVEngine::StringBatchWriteImpl(const WriteBatch::KV &kv,
...
 
     // Deleting kv is not existing
-    if (kv.type == StringDeleteRecord) {
-      if (found) {
-        batch_hint.pmem_record_to_free = hash_entry.index.string_record;
-        entry_ptr->Clear();
-      }
+    if (kv.type == StringDeleteRecord && !found) {
+      batch_hint.space_not_used = true;
+      persistSpaceEntry(batch_hint.allocated_space);
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1424]: markEmptySpace()?

File: engine/kv_engine.cpp [hunk 10]
```diff
@@ -1610,11 +1866,11 @@ Status KVEngine::RestoreDlistRecords(DLRecord *pmp_record) {
...
       GlobalLogger.Error("Bad linkage!\n");
-      if (DEBUG_LEVEL == 0) {
-        GlobalLogger.Error(
-            "Bad linkage can only be repaired when DEBUG_LEVEL > 0!\n");
-        std::abort();
-      }
+#if DEBUG_LEVEL == 0
+      GlobalLogger.Error(
+          "Bad linkage can only be repaired when DEBUG_LEVEL > 0!\n");
+      std::abort();
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:1872]: Don't

File: engine/kv_engine.cpp [hunk 11]
```diff
@@ -1833,13 +2087,12 @@ Status KVEngine::RestoreQueueRecords(DLRecord *pmp_record) {
...
-      // Bad linkage handled by DlinkedList.
-      if (DEBUG_LEVEL == 0) {
-        GlobalLogger.Error(
-            "Bad linkage can only be repaired when DEBUG_LEVEL > 0!\n");
-        std::abort();
-      }
-
+// Bad linkage handled by DlinkedList.
+#if DEBUG_LEVEL == 0
+      GlobalLogger.Error(
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:2092]: Don't

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/version/old_records_cleaner.cpp [hunk 12]
```diff
@@ -0,0 +1,258 @@
...
+    if (cleaner_thread_cache.old_data_records.size() > 0) {
+      std::lock_guard<SpinMutex> lg(cleaner_thread_cache.old_records_lock);
+      global_old_data_records_.emplace_back();
+      global_old_data_records_.back().swap(
+          cleaner_thread_cache.old_data_records);
+    }
+  }
+
+  // Find free-able data records
+  uint64_t handled_cnt = 0;
```
Comment (ZiyanShi) [inline engine/version/old_records_cleaner.cpp:72]: Are these variables for debug only?

File: engine/version/old_records_cleaner.cpp [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+  handled_cnt = 0;
+  delayed_cnt = 0;
+  for (auto &delete_records : global_old_delete_records_) {
+    for (auto &record : delete_records) {
+      if (record.release_time <= oldest_snapshot_ts) {
+        space_pending.entries.emplace_back(purgeOldDeleteRecord(record));
+        handled_cnt++;
+
+      } else {
+        delete_record_refered.emplace_back(std::move(record));
```
Comment (ZiyanShi) [inline engine/version/old_records_cleaner.cpp:98]: That's now how std::move is supposed to be used.

**Review by ZiyanShi — CHANGES_REQUESTED** [review]

File: engine/version/version_controller.hpp [hunk 14]
```diff
@@ -0,0 +1,182 @@
...
+  }
+
+  TimeStampType OldestSnapshotTS() {
+    return empty() ? kMaxTimestamp : head_.prev->GetTimestamp();
+  }
+
+private:
+  bool empty() { return head_.prev == &head_; }
+
+  SnapshotImpl head_;
```
[the remark this answers is not in the document]
Comment (JiayuZzz) [reply engine/version/version_controller.hpp:58]: no, we need delete a node in-place

File: engine/kv_engine.cpp (hunk 9 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:1424]: ok

File: engine/version/old_records_cleaner.cpp (hunk 12 above)
Comment (JiayuZzz) [reply engine/version/old_records_cleaner.cpp:72]: removed

File: engine/version/old_records_cleaner.cpp [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+  }
+  default:
+    std::abort();
+  }
+}
+
+SpaceEntry OldRecordsCleaner::purgeOldDeleteRecord(
+    const OldDeleteRecord &old_delete_record) {
+  DataEntry *data_entry =
+      static_cast<DataEntry *>(old_delete_record.pmem_delete_record);
```
[the remark this answers is not in the document]
Comment (JiayuZzz) [reply engine/version/old_records_cleaner.cpp:202]: we do not know type of pmem_delete_record (sorted or string), it's a void*

File: engine/kv_engine.cpp (hunk 7 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:1104]: I think a tool class is better than tokens everywhere

File: engine/kv_engine.cpp (hunk 8 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:1301]: check by try_lock() + unlock() is too heavy

File: engine/version/version_controller.hpp [hunk 16]
```diff
@@ -0,0 +1,182 @@
...
+
+  // These two used to get current timestamp of the instance
+  // version_base_: The newest timestamp on instance closing last time
+  // tsc_on_startup_: The CPU tsc on instance start up
+  uint64_t base_timestamp_;
+  uint64_t tsc_on_startup_;
+};
+
+class CheckPoint {
+public:
```
[the remark this answers is not in the document]
Comment (JiayuZzz) [reply engine/version/version_controller.hpp:170]: Snapshot and checkpoint is two things, though they have similar struct

**Pull request merged**